File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
class Solution {
public:
    int trap(vector<int>& h) {
        vector<int> prefixMax(h.size(), 0), suffixMax(h.size(), 0);
        int ans = 0;
        for (int i = 1; i < h.size(); i++) {
            prefixMax[i] = max(prefixMax[i-1], h[i-1]);
        }
        for (int i = h.size() - 2; i > 0; i--) {
            suffixMax[i] = max(suffixMax[i+1], h[i+1]);
        }
        for (int i = 0; i < h.size(); i++) {
            ans += max(min(prefixMax[i], suffixMax[i]) - h[i], 0);
        }
        return ans;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-1.py
import math
class Solution:

    def computeNumHours(self, piles: List[int], k: int) -> int:
        ans = 0
        for p in piles:
            ans += math.ceil(p/k)
        return ans

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        maxBananaPile = max(piles)
        l = 1
        r = maxBananaPile
        ans = maxBananaPile

        while l <= r:
            mid = int(l + (r - l)/2)
            hrs = self.computeNumHours(piles, mid)
            if hrs <= h:
                ans = min(ans, mid)
                r = mid - 1
            else:
                l = mid + 1
        return ans
